src/renderers/shared/fiber/ReactFiberBeginWork.js

```diff
@@ -277,6 +277,11 @@ module.exports = function<T, P, I, TI, PI, C, CX, PL>(
     markRef(current, workInProgress);
 
     if (!shouldUpdate) {
+      // Context providers should defer to sCU for rendering
+      if (hasContext) {
+        invalidateContextProvider(workInProgress, false);
```

Comment:
The difference is whether functional components can terminate rendering. They can't with this PR if the context provider rerendered above them. Basically sCU on a functional component becomes useless in many cases.